boltons/ioutils.py

```diff
@@ -0,0 +1,322 @@
+
+    Example::
+
+        >>> from boltons import ioutils
+        >>> with ioutils.SpooledBytesIO() as f:
+        ...     f.write(b"Happy IO")
+        ...     _ = f.seek(0)
+        ...     isinstance(f.getvalue(), ioutils.binary_type)
+        True
+    """
+
+    def read(self, n=-1):
+        return self.buffer.read(n)
+
+    def write(self, s):
+        if not isinstance(s, binary_type):
+            raise TypeError("{0} expected, got {1}".format(
+                binary_type.__name__,
+                type(s).__name__
+            ))
+
+        self.buffer.write(s)
+        if self.tell() >= self._max_size:
+            self.rollover()
+
+    def readline(self, length=None):
+        return self.buffer.readline(length)
+
+    def readlines(self, sizehint=0):
+        return self.buffer.readlines(sizehint)
+
+    def rollover(self):
+        """Roll the StringIO over to a TempFile"""
+        if not self._rolled:
+            tmp = TemporaryFile()
+            pos = self._file.tell()
+            self.__buffer.seek(0)
+            tmp.write(self.__buffer.read())
+            tmp.seek(pos)
+            self.__buffer.close()
+            self.__buffer = tmp
+
+    @property
+    def _rolled(self):
+        return not isinstance(self.buffer, BytesIO)
+
+    @property
+    def buffer(self):
+        try:
+            return self.__buffer
+        except AttributeError:
+            self.__buffer = BytesIO()
+        return self.__buffer
+
+
+class SpooledStringIO(SpooledIOBase):
+    """
+    SpooledStringIO is a spooled file-like-object that only accepts unicode
+    values. On Python 2.x this means the 'unicode' type and on Python 3.x this
+    means the 'str' type. Values are accepted as unicode and then coerced into
+    utf-8 encoded bytes for storage. On retrieval, the values are returned as
+    unicode.
+
+    Example::
+
+        >>> from boltons import ioutils
+        >>> with ioutils.SpooledStringIO() as f:
+        ...     f.write(u"\u2014 Hey, an emdash!")
+        ...     _ = f.seek(0)
+        ...     isinstance(f.read(), ioutils.text_type)
+        True
+
+     """
+
+    def read(self, n=-1):
+        return self.buffer.read(n).decode('utf-8')
+
+    def write(self, s):
+        if not isinstance(s, text_type):
+            raise TypeError("{0} expected, got {1}".format(
+                text_type.__name__,
+                type(s).__name__
+            ))
+
+        self.buffer.write(s.encode('utf-8'))
+        if self.tell() >= self._max_size:
+            self.rollover()
+
+    def readline(self, length=None):
+        return self.buffer.readline(length).decode('utf-8')
+
+    def readlines(self, sizehint=0):
+        return [x.decode('utf-8') for x in self.buffer.readlines(sizehint)]
+
+    @property
+    def buffer(self):
+        try:
+            return self.__buffer
+        except AttributeError:
+            self.__buffer = EncodedFile(BytesIO(), data_encoding='utf-8')
```

Comment:
@markrwilliams  I don't think we want seek and read to act on bytes (that's really what EncodedFile actually buys). If you drop out that, then seek and read don't work the way they do on StringIO and that prevents this from being a useful drop in replacement. I don't know exactly how to make tell() not provide info about what byte you're on (and maybe it shouldn't).

Really though I think in retrospect, len on SpooledStringIO should report the number of codepoints and not the number of bytes (otherwise again we're off of the StringIO api). One of the reasons that these classes are useful is that they provide a consistent API for different kinds of file and file-like objects. 

I have an idea for making len work as expected. I'll think about tell a bit more.